### PR TITLE
lightworks: init at 14.0.0

### PR DESCRIPTION
**Pull request by antonxy**
###### Motivation for this change
add Lightworks (Commercial Video Editor)

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This is my first PR so please let me know if anything does not fit.
Especially: Is it ok to add --enable-cxx in portaudio? According to nox-review 40 packages are affected

### CODE REVIEW

File: pkgs/development/libraries/portaudio/default.nix [hunk 1]
```diff
@@ -12,7 +12,7 @@ stdenv.mkDerivation rec {
   buildInputs = [ pkgconfig ]
     ++ stdenv.lib.optional (!stdenv.isDarwin) alsaLib;
 
-  configureFlags = [ "--disable-mac-universal" ];
+  configureFlags = [ "--disable-mac-universal --enable-cxx" ];
```
Comment (bjornfor) [inline pkgs/development/libraries/portaudio/default.nix]: This change should go in a separate commit.

File: pkgs/applications/video/lightworks/default.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
+{ stdenv, fetchurl, dpkg, makeWrapper, patchelf, buildFHSUserEnv,
+  gtk3, gnome3, gdk_pixbuf, cairo, libjpeg_original, glib, gnome2, mesa_glu, nvidia_cg_toolkit, zlib, openssl, portaudio }:
```
Comment (bjornfor) [inline pkgs/applications/video/lightworks/default.nix]: Please break overlong lines (>80 chars) like this:

```
{ stdenv ...
, gtk3 ...
}:
```

Note the placement of comma.

**Comment by antonxy** [comment]
Thanks for the review, should be fixed

File: pkgs/applications/video/lightworks/default.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      cp -r usr/share $out/share
+    '';
+
+    dontPatchELF = true;
+
+    meta = {
+      description = "Professional Non-Linear Video Editor";
+      homepage = "https://www.lwks.com/";
+      license = stdenv.lib.licenses.unfree;
+      maintainers = [ stdenv.lib.maintainers.antonxy ];
```
Comment (Mic92) [inline pkgs/applications/video/lightworks/default.nix]: Please limit platforms to `linux-x86_64` here.

File: pkgs/applications/video/lightworks/default.nix [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      ${patchelf}/bin/patchelf \
+        --interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+        $out/lib/lightworks/spawn
+
+       wrapProgram $out/lib/lightworks/ntcardvt \
+        --prefix LD_LIBRARY_PATH : ${fullPath}:$out/lib/lightworks \
+        --set FONTCONFIG_FILE $out/lib/lightworks/fonts.conf
+
+       wrapProgram $out/lib/lightworks/spawn \
+        --prefix LD_LIBRARY_PATH : ${fullPath}:$out/lib/lightworks
```
Comment (Mic92) [inline pkgs/applications/video/lightworks/default.nix]: this wrapper can be replaced by using the `rpath` option of patchelf:

```console
 patchelf \
    --interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
    --set-rpath "$out/lib/lightworks:${fullPath}" \
    $out/lib/lightworks/spawn
```

File: pkgs/applications/video/lightworks/default.nix [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      # This adds it to lightworks' search path while keeping the default
+      # using the FONTCONFIG_FILE env variable
+      echo "<?xml version='1.0'?>
+      <!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
+      <fontconfig>
+          <dir>/usr/share/fonts/truetype</dir>
+          <include>/etc/fonts/fonts.conf</include>
+      </fontconfig>" > $out/lib/lightworks/fonts.conf
+
+      ${patchelf}/bin/patchelf \
```
Comment (Mic92) [inline pkgs/applications/video/lightworks/default.nix]: You can just use `patchelf` directly as it is in the search path of `stdenv`

File: pkgs/applications/video/lightworks/default.nix [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    buildInputs = [ dpkg makeWrapper ];
+
+    phases = [ "unpackPhase" "installPhase" ];
+    unpackPhase = "dpkg-deb -x ${src} ./";
+
+    installPhase = ''
+      mkdir -p $out/bin
+      cp usr/bin/lightworks $out/bin
+      sed -i "s|/usr/lib/lightworks|$out/lib/lightworks|g" $out/bin/lightworks
```
Comment (Mic92) [inline pkgs/applications/video/lightworks/default.nix]:
```
cp usr/bin/lightworks $out/bin
sed -i "s|/usr/lib/lightworks|$out/lib/lightworks|g" $out/bin/lightworks
```

can be replaced by a single:

```
substitute usr/bin/lightworks $out/bin/lightworks \
   --replace "/usr/lib/lightworks" "$out/lib/lightworks" 
```

**Comment by antonxy** [comment]
Thanks, I updated that.
The `--set-rpath` option did not seem to work, the program did not find the libraries in the `$out/lib/lightworks` folder.
But I noticed that the `spawn` binary is not used anyway, so I removed it (it spawns `ntcardvt` and displays a loading bar, but the `lightworks` script in `bin` calls `ntcardvt` directly).

**Comment by 7c6f434c** [comment]
@GrahamcOfBorg sample 3

**Comment by rfhayashi** [comment]
@antonxy, I'm getting `/nix/store/z68sscqishcgpag75y8l0n0bi8ghcfk9-lightworks-14.0.0/lib/lightworks/ntcardvt: /nix/store/fapd1dnr70w57sd7lp5zzmzml6km9037-openssl-1.0.2n/lib/libcrypto.so.1.0.0: no version information available (required by /nix/store/z68sscqishcgpag75y8l0n0bi8ghcfk9-lightworks-14.0.0/lib/lightworks/libOSPrivate.so)` when running lightworks using this package. Do you know how to fix it? Let me know if I should report that somewhere else.

**Comment by antonxy** [comment]
Hm, I can reproduce the crash. It does not crash on the last commit before the merge, but crashes on the commit of the merge. But the "no version information available" message also appears on the working build so that does not seem to be the issue.

`strace -f -s 900 lightworks` outputs somwhere near the crash
```
write(2, "/nix/store/gbdjahb5rwbl45a18xpqrq6z83c4zhgh-lightworks-14.0.0/lib/lightworks/ntcardvt: symbol lookup error: /nix/store/gbdjahb5rwbl45a18xpqrq6z83c4zhgh-lightworks-14.0.0/lib/lightworks/PortAudio.lwd: undefined symbol: _ZN9portaudio6System14DeviceIteratorneERKS1_\n", 263) = 263
```
so it seems there has been some change in portaudio causing this.

Indeed if I override portaudio with the version from the working commit lightworks starts again.

The reason seems to be that `System::DeviceIterator::operator!= ( System::DeviceIterator const& rhs )` was made `const`.

https://abi-laboratory.pro/tracker/compat_report/portaudio/19_20140130/190600_20161030/b4bea/abi_compat_report.html

The best solution I can think of would be to create a new package of portaudio for the old version and use that until a newer version of lightworks is released which will hopefully use the new version of portaudio. I will prepare a pull request for that.
